grammar.peg

```diff
@@ -103,14 +103,29 @@ MatchNotMatches <- _ "not" _ "matches" _ {
    return MatchNotMatches, nil
 }
 
-Selector "selector" <- first:Identifier rest:SelectorOrIndex* {
-   sel := Selector{
-      first.(string),
+Selector <- slash:"/"? first:Identifier rest:SelectorOrIndex* {
```

Comment:
I think we may need a separate rule here to specifically match json pointer type things including escapes.

In particular adding "/" splitting support to the `SelectorOrIndex` rule a few lines below would cause the grammar to accept something like `first.second/third/fourth.fifth`

Similarly I think that requiring identifiers for the json pointer path parts is more restrictive than the spec would want.

What we probably want is something like:


```
Selector "selector" <- first:Identifier rest:SelectorOrIndex* {
   <pre-existing code>
} / '"' ptr:JSONPointer '"' {
   return Selector{
      Type: SelectorTypeJsonPointer,
      Path: ptr,
   }, nil
}

JSONPointer "json-pointer" <- ("/" !"/"*)* {
   var parts []string
   for _, part := range strings.Split(string(c.text), "/") {
      part = strings.ReplaceAll(part, "~0", "~")
      part = strings.ReplaceAll(part, "~1", "/")
      parts = append(parts, part)
   }
   
   return parts, nil
}
```

Not sure if that is 100% syntactically correct but you can see the point. We allow JSON pointer expressions when quoted and allow pretty much any valid unicode chars between slashes and expand ~0 and ~1 as the spec requires.

Reply:
We could potentially allow an unquoted JSON pointer expression too but it would be more limited to not include chars like `=` `!` and whitespace to keep the grammar unambiguous.

Reply:
I know that this grammar allows mixed-type selectors, I didn't think that was necessarily a problem though. Although you're right about char sets, I hadn't yet gotten to thinking about Mitchell's case where he would actually need to do the escaping. I'll separate them out.